### ISSUE TITLE
Weird logic about generating subs in ireps_generator.py

### ISSUE THREAD

**Issue by hannes-steffenhagen-diffblue** [diffblue/gnat2goto#55]
See the large comment in ireps_generator.py in this commit: https://github.com/diffblue/gnat2goto/pull/53/commits/675c6ce86b246c8e67ca1586fe5773288e562b45

Basically, we have some strange situations in which we do have a non-empty list of subs for irep members... but we don't actually want to fill in any subs.

This happens for example for `floatbv_type.json` which is a child of `bitvector_type.json` which itself is a child of `type_with_subtype.json` - i.e. `floatbv_type.json` is generated with a list of subtypes, but never actually has a meaningful subtype. It's possible that our specs don't really match up with what CBMC expects in a couple of cases like this.

**Comment by martin-cs** [comment]
IIRC ireps_generator.py was a test utility for the JSON_symtab front-end and semi / psuedo-independent of gnat2goto (they both use the same interface but nothing more).  Maybe it should live with the CPROVER patch.  @smowton may have opinions on the topic...

**Comment by smowton** [comment]
Nah, that code generates Ada data structures from the irep spec JSON files. It's pretty complicated though because Florian wanted to (a) share members wherever possible (using some fun auto-generated opportunistic union thing) and (b) have all ireps with common shape, so they can be allocated easily with a pool allocator.

Also I note in the meantime the irep hierarchy has changed a bit: (1) `bitvector_typet` no longer inherits `type_with_subtypet`, (2) `pointer_typet` has comments indicating it is both `bitvector_typet` and `type_with_subtypet`, but it actually isn't (the `subtype` accessor is still provided by `typet` itself).

**Comment by martin-cs** [comment]
I stand corrected.  @smowton where should this live and how should it be maintained?

**Comment by smowton** [comment]
I'd say it belongs in this repo as it generates Ada to be compiled along with the hand-written files here. The `spec` files ought ideally to be in cbmc/develop I reckon, but in that case we should have some way of verifying the correspondence between `std_types/expr/code` and the corresponding specs, or even generating part of the C++ headers similarly to how the Ada data structures are generated.
